bugwarrior/services/gitlab.py

```diff
@@ -70,6 +74,197 @@ def default_priorities(cls, values):
+            # work and instead return the entire result no matter what. Detect
+            # this by seeing if the results are the same as the last time
+            # around and bail if so. Unfortunately, while it is a GitLab bug,
+            # we have to deal with instances where it exists.
+            if items == detect_broken_gitlab_pagination:
+                break
+            detect_broken_gitlab_pagination = items
+
+            full += items
+
+            if len(items) < params['per_page']:
+                break
+            params['page'] += 1
+
+        return full
+
+    def get_repos(self, include_repos=None, include_regex=None,
+                  only_membership=False, only_owned=False):
+        """Returns a list of repo objects for all repositories accessible. Respects
+        config.include_repos"""
+
+        all_repos = []
+        if include_repos and not include_regex:
+            for repo in include_repos:
+                if repo.startswith("id:"):
+                    repo = repo[3:]
+                indiv_tmpl = 'projects' + '/' + quote(repo, '') + '?simple=true'
+                item = self._fetch(indiv_tmpl)
+                if not item:
+                    break
+
+                all_repos.append(item)
+
+        else:
+            querystring = {'simple': True}
+            if only_membership:
+                querystring['membership'] = True
+            if only_owned:
+                querystring['owned'] = True
+            all_repos = self._fetch_paged('projects' + '?' + urlencode(querystring))
+        return all_repos
+
+    def get_repo(self, repo_id: int) -> dict:
+        """Queries information about a single repository as JSON dictionary
+
+        :param repo_id: Project ID in the Gitlab instance
+        :type repo_id: int
+        :rtype: dict
+        """
+        return self._fetch('projects/' + str(repo_id))
+
+    def get_notes(self, rid: int, issue_type: str, issueid: int) -> list:
+        """Get notes attached to a certain issue / merge_request as list of JSON dictionaries
+
+        :param rid: Project ID in the Gitlab instance
+        :type rid: int
+        :param issue_type: "issues" / "merge_requests / snippets / epics"
+        :type issue_type: str
+        :param issueid: ID of issue
+        :type issueid: int
+        :rtype: list
+        """
+        return self._fetch_paged(f'projects/{rid}/{issue_type}/{issueid}/notes')
+
+    def get_repo_issues(self, rid: int) -> dict:
+        """Get all issues from a repository as JSON dictionary
+
+        :param rid: Project ID in the Gitlab instance
+        :type rid: int
+        :rtype: list
+        """
+        issues = {}
+        try:
+            repo_issues = self._fetch_paged('projects/%d/issues?state=opened' % rid)
+        except OSError:
+            # Projects may have issues disabled.
+            return {}
+        for issue in repo_issues:
+            issues[issue['id']] = (rid, issue)
+        return issues
```

Comment:
I'm still thinking there are opportunities to reduce duplication. For example, couldn't `get_repo_issues` be eliminated by simply passing `'projects/%d/issues?state=opened' % rid` into `get_issues_from_query`? (I suspect the structure of `get_repo_issues` is slightly more correct though.)

Reply:
Good point. I've change the calls accordingly.